File: 077-combinations-dfs.cpp
class Solution {
	public:
		std::vector<std::vector<int>> combine(int n, int k) {
			std::vector<std::vector<int>> result;
			dfs_(n, k, std::vector<int>(), 1, result);
			return result;
		}

	private:
		void dfs_(size_t n, size_t k, std::vector<int> array, size_t index, std::vector<std::vector<int>> &result) {
			if (k == array.size()) {
				result.push_back(array);

				return;
			}

			for (size_t i = index; i != n + 1; ++i) {
				array.push_back(i);
				dfs_(n, k, array, 1 + i, result);
				array.pop_back();
			}

		}
};

File: 077-combinations-dfs/077-combinations-dfs.rb
def dfs(n, k, array, index, result)
	if k == array.size
		result.push(array.dup)
		return
	end

	(index..n).each {|i|
		array.push(i)
		dfs(n, k, array, 1 + i, result)
		array.pop
	}
end

# @param {Integer} n
# @param {Integer} k
# @return {Integer[][]}
def combine(n, k)
	result = []
	dfs(n, k, [], 1, result)
	return result
end

p combine(2, 1)
